### ISSUE TITLE
[V3] Less abstraction in data management for larger datasets

### ISSUE THREAD

**Issue by ghost** [Cog-Creators/Red-DiscordBot#1239]
Config is a great tool for quick settings data: a name, a message id, or a boolean. It uses various abstract methods to save the data in a fast and easy to access way with the scopes it currently has.

This however does not apply for larger datasets such as comprehensive dictionaries that need frequent, and on the fly, data manipulation without immediately storing it away to the disk. At this point the abstract methods of accessing data through Config becomes inefficient for designing a cog and creates needless complexity for the programmer.

In V2, data is handled with dataIO, which works like a charm for the most part, but doesn't offer the same flexibility it has with smaller datasets, for which Config is seemingly designed. And, of course, the backend of Config is light seconds beyond dataIO in many regards.

Putting away whole data structures without the need to fit it into a scope gives an advantage for larger datasets. What I propose is giving Config methods to access and store complete data structures in a way dataIO did, but giving the flexible backend that Config offers. 

@mikeshardmind (Sinbad#0413) gave a well designed example of how to currently work around it, but it isn't part of Config yet.

This will work fine in most cases, but it would be really great to have this native feature in Config. And I know some people who would be pleased to have the same data manipulation flexibility as they had with dataIO.

```py
    def __init__(self, bot):
        self.bot = bot
        self.config = Config.get_conf(
            self, identifier=78631113035100160, force_registration=True)
        self.config.register_global(settings={})
        self.bot.loop.create_task(self.load_settings())
    
    async def load_settings(self):
        self.settings = await self.config.settings()

    async def save_settings(self):
        await self.config.settings.set(self.settings)```

**Comment by Tobotimus** [comment]
Great description of the issue, however I think more discussion is needed on the solution.

As you say, Config is great for (and mostly directed at) small datasets such as configurations for the whole bot or individual guilds/users etc. (hence the name, Config). For this reason I think this alternative data management system being suggested could be entirely separate from Config. This way, the cog could have complete control over the actual data structure as it is stored, and also when the data is saved.

**Comment by mikeshardmind** [comment]
Technically speaking, Cog creators already can do whatever with their data, while still being compliant with Red's storage model (see below.) However, I agree that it would be nice to be able to store other things with config, even if it just means sticking functions like the above or below into Config.

How best to go about this is another issue. Currently, supporting non-flat data is easy enough. However, making this a part of config will make writing drivers for certain other storage options (Redis comes to mind here) significantly more challenging.

```py
import redbot.core.data_manager as datam
import json

    def __init__(self, bot):
        self.path = datam.cog_data_path(self)

    self.save_json(self, filename, thing):
        with open(self.path + '/{}.json'.format(filename),encoding='utf-8', mode="w") as f:
            json.dump(data, f, indent=4,sort_keys=True, separators=(',',' : '))
        return data

    self.load_json(self, filename):
        with open(self.path + '/{}.json'.format(filename), encoding='utf-8', mode="r") as f:
            data = json.load(f)
        return data
```

**Comment by Redjumpman** [comment]
I have to agree with Paddo about the difficulties faced when you have a much more complicated data structure. Config has a lot of really great tools and the functionality is definitely much more robust than DataIO. However, it feels really cumbersome to write and not completely intuitive. It feels like I am importing a Python binding for a Java API.

In addition, some of the best methods are only in the API reference and not in the tutorial. For example, `all()`, `get_attr()`, and `set_attr()` are only in the API reference.  With my data structures I *have* to use `get_attr()` and `set_attr()`, but the API reference says `Use of this method should be avoided wherever possible.`, but doesn't explain why. What's the problem with dynamic attribute access?

I second what @PaddoInWonderland hinted towards. An adapter class to streamline some of these methods would be great. Also, I think the docs should have more examples for the other methods and some practical tutorials for complex structures.

**Comment by mikeshardmind** [comment]
Worth mentioning that the snippet of mine @PaddoInWonderland quoted doesn't even work, because behind the scenes, Config silently turns that into a Config Group object. the below would work, but the lengths required to work around config rather than with it here, should demonstrate that there needs to be another way to interact with config for more complicated data sets.
```py
    def __init__(self, bot):
        self.bot = bot
        self.config = Config.get_conf(
            self, identifier=78631113035100160, force_registration=True)
        self.config.register_global(settings=None)
        self.bot.loop.create_task(self.load_settings())
    
    async def load_settings(self):
        settings = await self.config.settings()
        if settings = None:
            self.settings = {}
        else:
            self.settings = json.loads(self.settings)

    async def save_settings(self):
        await self.config.settings.set(json.dumps(self.settings))
```

**Comment by mikeshardmind** [comment]
I think something that would help here is a method in config to retrieve all of a cog's data at once and a second method to write multiple (but not necessarily) all entries at once. 

Both of these methods would need to work with the entirety of the data, not just a specific config group.

**Comment by tekulvw** [comment]
#1335 Should have provided a lot more flexibility to Config by allowing you to specify your own groups and giving you access to the data at the driver level using `get_raw` and `set_raw` which exist on all `Group` objects. The documentation provided in #1342 will help demonstrate a small set of uses cases for these added features.

**Comment by Tobotimus** [comment]
Is this resolved now that #1335 and #1342 are merged?

**Comment by tekulvw** [comment]
Closing due to lack of response and because we believe the issues that Paddo brought up were sufficiently covered in #1335 and #1342 and successive config PRs.
